### PR TITLE
Refactor [Codecov]

### PR DESCRIPTION
**Pull request by calebcartwright**
This is ready for review now 

Couple notes (more detailed versions below):
- Token (for private repos) moved from route param to query param via new redirector functionality
- The coverage value is now rounded to the nearest whole number to be consistent with our other coverage badges (the legacy service always rounded down before, as does the native Codecov badge)

~~Still a work in progress, but opening this as a draft PR to start a dialog on a few items~~

- **the `token` parameter**
  The legacy service supported coverage badges for private repos via a user-supplied token in a route param. Typical/public repo pattern was `codecov/c/:vcs/:user/:repo`, for private repos the user adds `/token/:token` before the vcs param: `codecov/c/token/:token/:vcs/:user/:repo`. I was thinking it would be easier if the token was supplied via a query param (`codecov/c/:vcs/:user/:repo?token=`) vs. trying to keep the optional token in the route and having to use`format/capture`. I figured I could then create a redirect service for the token route to maintain backwards compatibility. Thoughts?
- **codecov api**
  The legacy service was using Codecov's native badge (https://codecov.io/gh/codecov/example-python/graphs/badge.svg) with plain text (https://codecov.io/gh/codecov/example-python/graphs/badge.txt). Codecov has a documented REST API for the data we need (https://docs.codecov.io/reference#section-get-a-single-repository) that returns a friendly JSON object, so I've switched to that. Any concerns there?
- **coverage rounding**
     The Codecov UI displays the exact coverage value, but their badges display coverage rounded down to the integer value (90.98% coverage shows as 90%), and that integer value is what our legacy service implementation saw and rendered. The Codecov API returns the precise coverage value (i.e. the 90.98%) so we've now got the option to either continue the Math.floor approach and be consistent with the native Codecov badges, or we could apply the same rounding that we do on other Shields coverage badges (90.98% would show as 91%) *I went with the latter in the refactored service, but let me know if anyone thinks we should switch back to rounding down*

~~Note: I've intentionally left the legacy service content in place (renamed) while this PR is in draft mode for easier side-by-side comparisons, but of course will remove it~~

### CODE REVIEW

**Comment by calebcartwright** [comment]
I'm having a little difficulty with the redirect around inserting the token query param _after_ the extension. 

I.e. we want to redirect the old path:
`/codecov/c/token/abc123def456/gh/codecov/private-example.svg`

to:
`/codecov/c/gh/codecov/private-example.svg?token=abc123def456` 

but right now my redirect is currently being sent to: 
`/codecov/c/gh/codecov/private-example?token=abc123def456.svg` 

Is there a way in the redirect service definition target to specify query params?

**Comment by paulmelnikow** [comment]
It definitely doesn't do that yet! 😁

I'd probably suggest adding a second function like `targetQueryParams`. Or maybe `target` should be renamed to `transformPath`, and `transformQueryParams` added.

I have a couple changes to `redirector.js` in flight right now for some work I'm doing related to #2068. Let me get to a pausing point and put a small bit of that into PR.

**Comment by calebcartwright** [comment]
> I'd probably suggest adding a second function like `targetQueryParams`. Or maybe `target` should be renamed to `transformPath`, and `transformQueryParams` added.

That's exactly what I was thinking too! I had started with renaming `target` to `targetUrl` and added `addTargetQueryParams` but I like what you've described here much better. There's nothing urgent about the Codecov refactor here whatsoever so happy to hold off on this one. 

Let me know if I can be of any help with the `redirector` stuff, otherwise I'll just keep an eye out and come back to this one whenever we're ready 👍

**Comment by paulmelnikow** [comment]
So… that PR got big (#3093). 😝

If you want I could pick out the redirector validation bits, which probably could be reviewed separately, much quicker.

File: services/codecov/codecov.spec.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+'use strict'
+
+const { expect } = require('chai')
```
Comment (calebcartwright) [inline services/codecov/codecov.spec.js]: I'm going to look into leveraging `sazerac` here instead

**Comment by calebcartwright** [comment]
> If you want I could pick out the redirector validation bits, which probably could be reviewed separately, much quicker

I'd say don't worry about. Happy to wait til #3093 gets merged

File: services/codecov/codecov.service.js [hunk 2]
```diff
@@ -46,63 +76,58 @@ module.exports = class Codecov extends LegacyService {
           repo: 'example-python',
           branch: 'master',
         },
-        staticPreview: { label: 'coverage', message: '90%', color: 'green' },
+        staticPreview: this.render({ coverage: 90 }),
       },
       {
-        title: 'Codecov private',
-        pattern: 'token/:token/:vcsName/:user/:repo',
+        title: 'Codecov private repository',
```
Comment (calebcartwright) [inline services/codecov/codecov.service.js]: This doesn't actually have to be a separate example anymore does it? I believe I can/should add the token query param to the two above examples instead

Comment (paulmelnikow) [reply services/codecov/codecov.service.js]: Yup, that makes sense to me. Some of the other ones like this, like the npm custom registry examples, probably should be converted in the same way.

**Comment by calebcartwright** [comment]
I'm going to take a stab at updating the redirector function to support transforming query params

File: services/codecov/codecov.service.js [hunk 3]
```diff
@@ -1,108 +1,139 @@
...
+  }
+
   static get route() {
     return {
       base: 'codecov/c',
-      pattern: '',
+      // https://docs.codecov.io/docs#section-common-questions
+      // Github, BitBucket, and GitLab are the only supported options (long or short form)
+      pattern:
+        ':vcsName(github|gh|bitbucket|bb|gl|gitlab)/:user/:repo/:branch*',
```
Comment (calebcartwright) [inline services/codecov/codecov.service.js]: I was torn on whether to include both the long and shorthand value for the `vcsName`. The legacy service regex would let either through, so I felt that we'd need to support both long/short forms to avoid breaking any existing users. 

If we prefer one form vs. the other, then I could probably set up a few redirectors to maintain backwards compatibility with the other form. However, I elected to stick with the broader route param support here since there's already a redirector transforming the legacy token route parts to a query param.

Let me know what folks think

Comment (paulmelnikow) [reply services/codecov/codecov.service.js]: I'm inclined to standardize on the longer forms at some point, though I don't think it's important to create more redirectors right now.

File: services/codecov/codecov.service.js [hunk 4]
```diff
@@ -1,108 +1,139 @@
...
+  token: Joi.string(),
+}).required()
+
+const documentation = `
+  <p>
+    You may specify a Codecov token to get coverage for a private repository.
+  </p>
+  <p>
+    See the <a href="https://docs.codecov.io/reference#authorization">Codecov Docs</a>
+    for more information about creating a token.
```
Comment (paulmelnikow) [inline services/codecov/codecov.service.js]: As an aside, do you know if these are read-only / status-only tokens? I hate to remove wanted functionality, though at some point we really should lock down the remaining endpoints that accept read-write tokens. For the purpose of this PR preserving the behavior is 👍

File: services/codecov/codecov.service.js (hunk 3 above)
Comment (calebcartwright) [reply services/codecov/codecov.service.js]: 👍

File: services/codecov/codecov.spec.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+const { test, given } = require('sazerac')
+const Codecov = require('./codecov.service')
+
+const unknownCoverage = { coverage: 'unknown' }
+
+describe('Codecov', function() {
+  test(Codecov.prototype.transform, () => {
+    given({ json: {} }).expect(unknownCoverage)
+    given({ json: { commit: {} } }).expect(unknownCoverage)
+  })
```
Comment (paulmelnikow) [inline services/codecov/codecov.spec.js]: `sazerac` has a cool thing called `forCases` you can use here:

```js
test(Codecov.prototype.transform, () => {
  forCases([
    given({ json: {} }),
    given({ json: { commit: {} } }),
  ]).expect({ coverage: 'unknown' })
}
```

File: services/codecov/codecov.tester.js [hunk 6]
```diff
@@ -1,23 +1,59 @@
...
+  .get('/gh/codecov/private-example-python.json')
+  .intercept(nock =>
+    nock('https://codecov.io/api')
+      .get('/gh/codecov/private-example-python')
+      .reply(401)
+  )
+  .expectBadge({
+    label: 'coverage',
+    message: 'not authorized to access repository',
+  })
```
Comment (paulmelnikow) [inline services/codecov/codecov.tester.js]: This test is arguably helpful to ensure that things are wired up correctly, though it feels like it's mostly a test of `errorMessages`.

Could it use a real private project instead?

Comment (calebcartwright) [reply services/codecov/codecov.tester.js]: Fair point! I don't actually know of any private projects that have Codecov enabled, but if anyone has such a repo and is willing to share I'd be happy to plug it in.

File: services/codecov/codecov.spec.js (hunk 5 above)
Comment (calebcartwright) [reply services/codecov/codecov.spec.js]: 👍

File: services/codecov/codecov.service.js (hunk 4 above)
Comment (calebcartwright) [reply services/codecov/codecov.service.js]: I don't know, and it wasn't clear to me from the docs whether there were different permission scopes. My only experience with Codecov tokens has been repository-specific tokens that have permissions to publish coverage results

File: services/codecov/codecov-redirect.service.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    category: 'coverage',
+    route: {
+      base: 'codecov/c',
+      pattern:
+        'token/:token/:vcsName(github|gh|bitbucket|bb|gl|gitlab)/:user/:repo/:branch*',
+    },
+    transformPath: ({ vcsName, user, repo, branch }) =>
+      `/codecov/c/${vcsName}/${user}/${repo}${branch ? `/${branch}` : ''}`,
+    transformQueryParams: ({ token }) => ({ token }),
+    dateAdded: new Date('2019-02-23'),
```
[the remark this answers is not in the document]
Comment (calebcartwright) [reply services/codecov/codecov-redirect.service.js]: 😄 Good catch. I plugged that in before I realized the need for the redirector to be able to transform query params

File: services/codecov/codecov-redirect.service.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+module.exports = [
+  redirector({
+    category: 'coverage',
+    route: {
+      base: 'codecov/c',
+      pattern:
+        'token/:token/:vcsName(github|gh|bitbucket|bb|gl|gitlab)/:user/:repo/:branch*',
+    },
+    transformPath: ({ vcsName, user, repo, branch }) =>
+      `/codecov/c/${vcsName}/${user}/${repo}${branch ? `/${branch}` : ''}`,
```
[the remark this answers is not in the document]
Comment (calebcartwright) [reply services/codecov/codecov-redirect.service.js]: Long form?

File: services/codecov/codecov.service.js [hunk 9]
```diff
@@ -1,108 +1,139 @@
...
-// https://github.com/badges/shields/blob/master/doc/rewriting-services.md
-//
-// Do not base new services on this code.
-module.exports = class Codecov extends LegacyService {
+// https://docs.codecov.io/reference#totals
+// A new repository that's been added but never had any coverage reports
+// uploaded will not have a `commit` object in the response and some times
+// the `totals` object can also be missing for the latest commit.
+// Keeping these as optional in the schema and handling it in the transform
+// function to maintain consistent badge behavior with the legacy service implementation.
```
[the remark this answers is not in the document]
Comment (calebcartwright) [reply services/codecov/codecov.service.js]: Working on updating this. Initially this was another one of those "because that's the way it was done before in the legacy service" things 😄 

However, I think the underlying scenarios that the legacy service implementation was trying to support was valid (non error) user cases when the coverage was not yet known. 

I can try to dig into it a bit more

File: services/codecov/codecov.tester.js (hunk 6 above)
Comment (paulmelnikow) [reply services/codecov/codecov.tester.js]: I don't know one… Maybe we could just leave a comment.

File: services/codecov/codecov.service.js (hunk 4 above)
Comment (calebcartwright) [reply services/codecov/codecov.service.js]: I don't think their anchors are working, but:

https://codecov.io/site/security#how-does-codecov-authenticate-access-to-a-repository
https://codecov.io/site/security#what-scope-is-used-when-signing-up-using-github 

TL;DR if I'm reading correctly (at least for GH), scope carries through to Codecov perms, based on the token-owners permissions on the repo. So a user with read-only access to a private GH repo would create a token that would only be able to be used to see/read Codecov coverage info for that repo. A user with read/write access to that private GH repo would create a token that could both see/read coverage info, and publish/upload coverage results for that repo in Codecov.

File: services/codecov/codecov.tester.js (hunk 6 above)
Comment (calebcartwright) [reply services/codecov/codecov.tester.js]: Works for me!

File: services/codecov/codecov.service.js (hunk 4 above)
Comment (paulmelnikow) [reply services/codecov/codecov.service.js]: Sounds like it is _possible_ that folks might create separate account that enable this to be done securely, though it's also possible they won't have…

**Comment by paulmelnikow** [comment]
@calebcartwright Let me know if you want anything else from me before merging!

**Comment by calebcartwright** [comment]
@paulmelnikow sorry I should've left a comment. I want to spend a bit more time challenging the schema in relation to the API response as I thought of a couple other scenarios. Will also add a couple of the code comments we've discussed in this thread

Will give you a shout when I'm done

**Comment by calebcartwright** [comment]
Alrighty, I created some sample apps and ran through some various scenarios and concluded the schema is best as-is 👍 

@paulmelnikow - I'm all set here now
